sdks/go/pkg/beam/external.go

```diff
@@ -16,10 +16,151 @@
+	if !s.IsValid() {
+		// return nil, errors.New("invalid scope")
+		fmt.Println("invalid scope")
+	}
+	for i, col := range e.In {
+		if !col.IsValid() {
+			// return nil, errors.Errorf("invalid pcollection to external: index %v", i)
+			fmt.Printf("\ninvalid pcollection to external: index %v", i)
+
+		}
+	}
+
+	// Using exisiting MultiEdge format to represent ExternalTransform (already backwards compatible)
+	payload := &graph.Payload{
+		URN:  e.Urn,
+		Data: e.Payload,
+	}
+	var ins []*graph.Node
+	for _, col := range e.In {
+		ins = append(ins, col.n)
+	}
+	edge := graph.NewCrossLanguage(s.real, s.scope, ins, payload)
+
+	// TODO(pskevin): There needs to be a better way of associating this ExternalTransform to the pipeline
+	// Adding ExternalTransform to pipeline referenced by MultiEdge ID
+	if p.ExpandedTransforms == nil {
+		p.ExpandedTransforms = make(map[string]*ExternalTransform)
+	}
+	p.ExpandedTransforms[fmt.Sprintf("e%v", edge.ID())] = e
+
+	/*
+		Build the ExpansionRequest
+	*/
+	// Obtaining the components and transform proto representing this transform
+	pipeline, err := graphx.Marshal([]*graph.MultiEdge{edge}, &graphx.Options{})
+	if err != nil {
+		panic(err)
+	}
+
+	// Adding fake impulses to each input as required for correct expansion
+	// TODO(pskevin): Remove these fake impulses from final Pipeline since multiple producers of the same PCollections is logically wrong
+	transforms := pipeline.Components.Transforms
+	rootTransformID := pipeline.RootTransformIds[0]
+	for tag, id := range transforms[rootTransformID].Inputs {
+		key := fmt.Sprintf("%s_%s", "impulse", tag)
+
+		output := map[string]string{"out": id}
+		impulse := &pipepb.PTransform{
+			UniqueName: key,
+			Spec: &pipepb.FunctionSpec{
+				Urn: graphx.URNImpulse,
+			},
+			Outputs: output,
+		}
+
+		transforms[key] = impulse
+	}
+
+	// Assembling ExpansionRequest proto
+	req := &jobpb.ExpansionRequest{
+		Components: pipeline.Components,
+		Transform:  transforms[rootTransformID],
+		Namespace:  s.String(),
+	}
+
+	/*
+		Querying Expansion Service
+	*/
+	// Setting grpc client
+	conn, err := grpc.Dial(e.ExpansionAddr, grpc.WithInsecure())
+	if err != nil {
+		panic(err)
+	}
+	defer conn.Close()
+	client := jobpb.NewExpansionServiceClient(conn)
+
+	// Handling ExpansionResponse
+	res, err := client.Expand(context.Background(), req)
+	if err != nil {
+		panic(err)
+	}
+	e.Components = res.GetComponents()
+	e.ExpandedTransform = res.GetTransform()
+	e.Requirements = res.GetRequirements()
+
+	/*
+		Associating output PCollections of the expanded transform with correct internal outbound links and nodes
+	*/
+	// No information about the output types and bounded nature has been explicitly passed by the user
+	if len(e.Out) == 0 || cap(e.Out) == 0 {
+		// Infer output types from ExpansionResponse and update e.Out
+		if e.Out == nil {
+			// Use reverse schema encoding
+		} else {
+			// Use the coders list and map from coder id to internal FullType?
+		}
+	}
```

Comment:
Since the next change will handle PCollection mappings by user provided string keys, we can probably get rid of this for now. Do add a TODO(lostluck) to remind me to plug in the schema coders here instead of what we're doing for the output types.

Reply:
Makes sense. Will do!